### PR TITLE
New UI re-design for progress dialogs and basic alert dialogs

### PR DESCRIPTION
**Pull request by amstone326**
Redesign of progress dialogs and (basic) alert dialogs to comply with the new UI spec. Modeled after image in this ticket: http://manage.dimagi.com/default.asp?183307#1025658, and using the google design specs for dialogs: https://www.google.com/design/spec/components/dialogs.html#dialogs-specs. Screenshots included below! Have also tested on devices of a few different sizes to make sure it still looks right.

The new design has been applied to every progress dialog (by refactoring CustomProgressDialog), and nearly every alert dialog (by creating AlertDialogFactory). The alert dialogs that were excluded are those that involve more complicated construction in one way or another, and which Clayton therefore told me to leave out of the redesign, at least for now. 

The alert dialogs that were NOT included in the redesign - just so that we have a comprehensive list somewhere and in case anyone feels strongly that any of them should be included - are the following:
-buildAboutCommCareDialog() in DialogCreationHelpers
-onCreateDialog() in CallOutActivity
-showAboutCommCareDialog() and showDemoModeWarning() in CommCareHomeActivity
-createSortMenu() in EntitySelectActivity
-buildMenuList() in AndroidShortucts
-createQuitDrawDialog() in DrawActivity
-createLanguageDialog(), createQuitDialog(), and createRepeatDialog() in FormEntryActivity
-setupSplashPathPreference() in PreferencesActivity
-fireHelpText() in QuestionWidget
-the constructor of SpinnerMultiWidget

### CODE REVIEW

**Comment by biyeun** [comment]
@amstone326  screenshot?

**Comment by amstone326** [comment]
@biyeun, not ready to be looked at yet. I'll remove the "don't pull" label and ping you when it is.

**Comment by amstone326** [comment]
Progress dialog with a determinate progress bar:

screenshot_2015-10-12-13-40-43

**Comment by amstone326** [comment]
Progress dialog with an indeterminate progress bar:

screenshot_2015-10-12-13-41-50

**Comment by amstone326** [comment]
Alert dialog:

screenshot_2015-10-12-13-41-33

**Comment by amstone326** [comment]
Alert dialog with title icon:

screenshot_2015-10-12-15-30-42

File: app/src/org/commcare/dalvik/dialogs/DialogController.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
     /** Create an instance of CustomProgressDialog specific to the activity
      *  implementing this method -- this method should be implemented lower down
      *  in the activity hierarchy, in one of CommCareActivity's subclasses, 
-     *  while the other 3 methods can be handled entirely by CommCareActivity **/
```
Comment (amstone326) [inline app/src/org/commcare/dalvik/dialogs/DialogController.java]: fixing stale comment

File: app/src/org/odk/collect/android/activities/FormEntryActivity.java [hunk 2]
```diff
@@ -1348,42 +1346,38 @@ private void createQuitDialog() {
...
 
-                        mAlertDialog =
-                                new AlertDialog.Builder(this)
-                                        .setIcon(android.R.drawable.ic_dialog_info)
-                                        .setTitle(StringUtils.getStringRobust(this, R.string.quit_application, mFormController.getFormTitle()))
-                                                .setNeutralButton(StringUtils.getStringSpannableRobust(this, R.string.do_not_exit),
-                        new DialogInterface.OnClickListener() {
-                            @Override
-                            public void onClick(DialogInterface dialog, int id) {
+        AlertDialog dialog = new AlertDialog.Builder(this)
```
Comment (amstone326) [inline app/src/org/odk/collect/android/activities/FormEntryActivity.java]: Didn't apply redesign here, was just fixing unnecessary indenting

**Comment by amstone326** [comment]
@biyeun, @ctsims - Ready for review, screenshots included.

File: app/res/layout/custom_alert_dialog.xml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        android:layout_below="@id/dialog_message"
+        android:layout_toLeftOf="@id/neutral_button"
+
+        android:layout_marginBottom="@dimen/standard_spacer"
+        android:layout_marginRight="@dimen/standard_spacer"
+        android:layout_marginLeft="@dimen/standard_spacer"
+
+        android:background="@android:color/transparent"
+        android:textSize="@dimen/font_size_medium"
+        android:textColor="#2196F3"
```
Comment (ctsims) [inline app/res/layout/custom_alert_dialog.xml]: Can we make sure this is coming from the colors.xml file?

File: app/res/layout/custom_alert_dialog.xml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        android:padding="@dimen/standard_spacer"
+
+        android:layout_below="@id/dialog_message"
+        android:layout_toLeftOf="@id/positive_button"
+
+        android:layout_marginBottom="@dimen/standard_spacer"
+
+        android:background="@android:color/transparent"
+        android:textSize="@dimen/font_size_medium"
+        android:textColor="#2196F3"
```
Comment (ctsims) [inline app/res/layout/custom_alert_dialog.xml]: colors should be unified

File: app/res/values/dimens.xml [hunk 5]
```diff
@@ -21,6 +21,10 @@
     <dimen name="spacer_small">5dp</dimen>
     <dimen name="standard_spacer_half">4dp</dimen>
     <dimen name="standard_spacer">8dp</dimen>
+    <dimen name="standard_spacer_double">16dp</dimen>
+    <dimen name="standard_spacer_large">24dp</dimen>
+    <dimen name="standard_title_margin_bottom">20dp</dimen>
+    <dimen name="min_button_height">52dp</dimen>
```
Comment (ctsims) [inline app/res/values/dimens.xml]: These are from the material guidelines I imagine?

**Comment by ctsims** [comment]
@amstone326 Looks Great! I'm ready to pull after minor color resource change.

I noticed the same thing Phillip did regarding the custom dialog, namely when we launch it as "uncancellable" you still seem to be able to hit "Back" to dismiss it. I think we should address that before pulling, as this expands how many screens use the custom dialog

File: app/res/values/dimens.xml (hunk 5 above)
Comment (amstone326) [inline app/res/values/dimens.xml]: Yep

**Comment by amstone326** [comment]
@ctsims, @phillipm ..Really glad you mentioned that, I don't know how I missed that all this time. Turns out that calling setCancelable() on the dialog builder has no effect and you have to call it on the DialogFragment. Extremely annoying that that is the case, but fixed now. 

Color issue fixed as well, so should be mergeable.

File: app/res/layout/dialog_title.xml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    android:layout_marginTop="@dimen/standard_spacer_large"
+    android:layout_marginBottom="@dimen/standard_title_margin_bottom"
+    android:layout_marginRight="@dimen/standard_spacer_large"
+    android:layout_marginLeft="@dimen/standard_spacer_large" >
+
+    <ImageView
+        android:id="@+id/dialog_title_icon"
+        android:layout_height="wrap_content"
+        android:layout_width="wrap_content"
+        android:layout_alignParentTop="true"
```
Comment (phillipm) [inline app/res/layout/dialog_title.xml]: Linter says this param is invalid (and the instance below as well)

File: app/src/org/commcare/dalvik/dialogs/AlertDialogFactory.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import org.commcare.dalvik.R;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * Created by amstone326 on 10/9/15.
+ */
+public class AlertDialogFactory {
+
+    private AlertDialog dialog;
+    private View view;
```
Comment (phillipm) [inline app/src/org/commcare/dalvik/dialogs/AlertDialogFactory.java]: These can be made `final`
Also, I think your not using our custom code formatting configuration: https://confluence.dimagi.com/display/MD/Code+Conventions

**Comment by phillipm** [comment]
Did a quick pass and left minor comments, but looks good otherwise. I'm excited to have the dialogs updated to the new UI and that the cancel issue was resolve! Thanks for the screenshots and list of dialogs that have yet to be migrated.

File: app/src/org/commcare/dalvik/dialogs/AlertDialogFactory.java (hunk 7 above)
Comment (amstone326) [inline app/src/org/commcare/dalvik/dialogs/AlertDialogFactory.java]: Ah yea I don't :/ I tried setting them up once, ran into some road bumps, and didn't try again.. Will make sure to figure it out though.

**Comment by amstone326** [comment]
@ctsims Addressed Phillip's comments, so good to go

**Comment by biyeun** [comment]
Beautiful :heart_eyes:

**Pull request merged**